From a9c53900ec63a4380e041c3cf04c6a9d26a4ff09 Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Wed, 1 Oct 2014 17:40:34 -0400
Subject: [ethernet] cxgb4: Add warning msg when attaching to adapters which have FLASHes smaller than 2Mb

Message-id: <1412185238-11567-4-git-send-email-svemuri@redhat.com>
Patchwork-id: 96829
O-Subject: [RHEL7.1 PATCH BZ 1147683 3/7] cxgb4: Add warning msg when attaching to adapters which have FLASHes smaller than 2Mb
Bugzilla: 1147683
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>

Upstream : https://git.kernel.org/cgit/linux/kernel/git/davem/net-next.git/commit/drivers/net/ethernet/chelsio?id=c290607e3e7b8d24eb90468180f4cf96b371d8d0

Based on original work by Casey Leedom <leedom@chelsio.com>

Signed-off-by: Hariprasad Shenai <hariprasad@chelsio.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit c290607e3e7b8d24eb90468180f4cf96b371d8d0)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/chelsio/cxgb4/t4_hw.c b/drivers/net/ethernet/chelsio/cxgb4/t4_hw.c
index 0501ea2..63fa976 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/t4_hw.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/t4_hw.c
@@ -3838,6 +3838,10 @@ static int get_flash_params(struct adapter *adap)
  adap->params.sf_size = 1 << info;
  adap->params.sf_fw_start =
   t4_read_reg(adap, CIM_BOOT_CFG) & BOOTADDR_MASK;
+
+ if (adap->params.sf_size < FLASH_MIN_SIZE)
+  dev_warn(adap->pdev_dev, "WARNING!!! FLASH size %#x < %#x!!!\n",
+    adap->params.sf_size, FLASH_MIN_SIZE);
  return 0;
 }
 
diff --git a/drivers/net/ethernet/chelsio/cxgb4/t4_hw.h b/drivers/net/ethernet/chelsio/cxgb4/t4_hw.h
index 71b799b..55c3a6a 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/t4_hw.h
+++ b/drivers/net/ethernet/chelsio/cxgb4/t4_hw.h
@@ -206,6 +206,12 @@ enum {
  FLASH_CFG_START = FLASH_START(FLASH_CFG_START_SEC),
  FLASH_CFG_MAX_SIZE = FLASH_MAX_SIZE(FLASH_CFG_NSECS),
 
+ /* We don't support FLASH devices which can't support the full
+  * standard set of sections which we need for normal
+  * operations.
+  */
+ FLASH_MIN_SIZE = FLASH_CFG_START + FLASH_CFG_MAX_SIZE,
+
  FLASH_FPGA_CFG_START_SEC = 15,
  FLASH_FPGA_CFG_START = FLASH_START(FLASH_FPGA_CFG_START_SEC),
 
-- 
1.7.1